### PR TITLE
[Feature] Add WordAccuracy for OCR Task

### PR DESCRIPTION
**Pull request by Harold-lkk**
Add WordAccuracy Metric
Adapt at https://github.com/open-mmlab/mmocr/pull/1728

### CODE REVIEW

**Review by C1rN09 — COMMENTED** [review]
Good! Also please add api docs

File: mmeval/metrics/word_accuracy.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        mode (str or list[str]): Options are:
+            - 'exact': Accuracy at word level.
+            - 'ignore_case': Accuracy at word level, ignoring letter
+              case.
+            - 'ignore_case_symbol': Accuracy at word level, ignoring
+              letter case and symbol. (Default metric for academic evaluation)
+            If mode is a list, then metrics in mode will be calculated
+            separately. Defaults to 'ignore_case_symbol'.
+        valid_symbol (str): Valid characters. Defaults to
+            '[^A-Z^a-z^0-9^\u4e00-\u9fa5]'.
```
Comment (C1rN09) [inline mmeval/metrics/word_accuracy.py:21]: Maybe should add some explanations on this. Is it required to be a regular expresion? What format is acceptable?

File: mmeval/metrics/word_accuracy.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+    Args:
+        mode (str or list[str]): Options are:
+            - 'exact': Accuracy at word level.
+            - 'ignore_case': Accuracy at word level, ignoring letter
+              case.
+            - 'ignore_case_symbol': Accuracy at word level, ignoring
+              letter case and symbol. (Default metric for academic evaluation)
+            If mode is a list, then metrics in mode will be calculated
+            separately. Defaults to 'ignore_case_symbol'.
```
Comment (C1rN09) [inline mmeval/metrics/word_accuracy.py:19]:
```suggestion
    Args:
        mode (str or list[str]): Options are:
        
            - 'exact': Accuracy at word level.
            - 'ignore_case': Accuracy at word level, ignoring letter
              case.
            - 'ignore_case_symbol': Accuracy at word level, ignoring
              letter case and symbol. (Default metric for academic evaluation)
              
            If mode is a list, then metrics in mode will be calculated
            separately. Defaults to 'ignore_case_symbol'.
```

File: mmeval/metrics/word_accuracy.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        valid_symbol (str): A regular expression to filter out invalid or
+            not cared characters. Defaults to '[^A-Z^a-z^0-9^\u4e00-\u9fa5]'.
```
Comment (C1rN09) [inline mmeval/metrics/word_accuracy.py:21]: Same issue with `\u4e00-\u9fa5` in api docs rendering.

File: mmeval/metrics/word_accuracy.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        Args:
+            results (list[float]): The processed results of each batch.
+
+        Returns:
+            dict[str, float]: Nested dicts as results. Provided keys are:
+            - accuracy (float): Accuracy at word level.
+            - ignore_case_accuracy (float): Accuracy at word level, ignoring
+                letter case.
+            - ignore_case_symbol_accuracy (float): Accuracy at word level,
+                ignoring letter case and symbol.
```
Comment (C1rN09) [inline mmeval/metrics/word_accuracy.py:87]:
```suggestion
            dict[str, float]: Nested dicts as results. Provided keys are:
            
            - accuracy (float): Accuracy at word level.
            - ignore_case_accuracy (float): Accuracy at word level, ignoring
              letter case.
            - ignore_case_symbol_accuracy (float): Accuracy at word level,
              ignoring letter case and symbol.
```

**Comment by Harold-lkk** [comment]
https://github.com/open-mmlab/mmocr/pull/1728

**Review by C1rN09 — APPROVED** [review]

File: mmeval/metrics/word_accuracy.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        super().__init__(**kwargs)
+        self.mode = mode
+        self.invalid_symbol = re.compile(invalid_symbol)
+        assert isinstance(mode, (str, list))
+        if isinstance(mode, str):
+            mode = [mode]
+        assert all([isinstance(item, str) for item in mode])
+        assert set(mode).issubset(
+            {'exact', 'ignore_case', 'ignore_case_symbol'})
+        self.mode = set(mode)  # type: ignore
```
Comment (zhouzaida) [inline mmeval/metrics/word_accuracy.py:52]:
```suggestion
        assert all(isinstance(item, str) for item in mode)
        self.mode = set(mode)  # type: ignore
        assert self.mode.issubset(
            {'exact', 'ignore_case', 'ignore_case_symbol'})
```

File: mmeval/metrics/word_accuracy.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            results (list[float]): The processed results of each batch.
+
+        Returns:
+            dict[str, float]: Nested dicts as results. Provided keys are:
+
+            - accuracy (float): Accuracy at word level.
+            - ignore_case_accuracy (float): Accuracy at word level, ignoring
+                letter case.
+            - ignore_case_symbol_accuracy (float): Accuracy at word level,
+                ignoring letter case and symbol.
```
Comment (zhouzaida) [inline mmeval/metrics/word_accuracy.py:90]:
```suggestion
            - ignore_case_accuracy (float): Accuracy at word level, ignoring
              letter case.
            - ignore_case_symbol_accuracy (float): Accuracy at word level,
              ignoring letter case and symbol.
```

File: mmeval/metrics/word_accuracy.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+            ignore_case_sum += ignore_case
+            ignore_case_symbol_sum += ignore_case_symbol
+        if 'exact' in self.mode:
+            eval_res['accuracy'] = exact_sum / gt_word_num
+        if 'ignore_case' in self.mode:
+            eval_res['ignore_case_accuracy'] = ignore_case_sum / gt_word_num
+        if 'ignore_case_symbol' in self.mode:
+            eval_res['ignore_case_symbol_accuracy'] =\
+                ignore_case_symbol_sum / gt_word_num
+        return eval_res
```
Comment (zhouzaida) [inline mmeval/metrics/word_accuracy.py:106]:
```suggestion
        metric_results = {}
        gt_word_num = max(len(results), 1.0)
        exact_sum, ignore_case_sum, ignore_case_symbol_sum = 0.0, 0.0, 0.0
        for exact, ignore_case, ignore_case_symbol in results:
            exact_sum += exact
            ignore_case_sum += ignore_case
            ignore_case_symbol_sum += ignore_case_symbol
        if 'exact' in self.mode:
            metric_results['accuracy'] = exact_sum / gt_word_num
        if 'ignore_case' in self.mode:
            metric_results['ignore_case_accuracy'] = ignore_case_sum / gt_word_num
        if 'ignore_case_symbol' in self.mode:
            metric_results['ignore_case_symbol_accuracy'] =\
                ignore_case_symbol_sum / gt_word_num
        return metric_results
```

File: mmeval/metrics/word_accuracy.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+        self.invalid_symbol = re.compile(invalid_symbol)
+        assert isinstance(mode, (str, list))
+        if isinstance(mode, str):
+            mode = [mode]
+        assert all([isinstance(item, str) for item in mode])
+        assert set(mode).issubset(
+            {'exact', 'ignore_case', 'ignore_case_symbol'})
+        self.mode = set(mode)  # type: ignore
+
+    def add(self, predictions: Sequence[str], labels: Sequence[str]) -> None:  # type: ignore # yapf: disable # noqa: E501
```
Comment (zhouzaida) [inline mmeval/metrics/word_accuracy.py:54]:
```suggestion
    def add(self, predictions: Sequence[str], groundtruths: Sequence[str]) -> None:  # type: ignore # yapf: disable # noqa: E501
```

File: mmeval/metrics/word_accuracy.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+            - 'ignore_case_symbol': Accuracy at word level, ignoring
+              letter case and symbol. (Default metric for academic evaluation)
+
+            If mode is a list, then metrics in mode will be calculated
+            separately. Defaults to 'ignore_case_symbol'.
+        invalid_symbol (str): A regular expression to filter out invalid or
+            not cared characters. Defaults to '[^A-Za-z0-9\u4e00-\u9fa5]'
+        **kwargs: Keyword parameters passed to :class:`BaseMetric`.
+
+    Example:
```
Comment (zhouzaida) [inline mmeval/metrics/word_accuracy.py:26]:
```suggestion
    Examples:
```

**Review by zhouzaida — APPROVED** [review]

**Pull request merged**